src/v/lsm/block/builder.cc

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "lsm/block/builder.h"
+
+#include <limits>
+#include <utility>
+
+namespace lsm::block {
+
+builder::builder()
+  : builder(builder::options{}) {}
+
+builder::builder(builder::options opts) {
+    _restarts.reserve(opts.restart_interval);
+    // First restart point is at offset 0
+    _restarts.push_back(0);
+}
+
+size_t builder::current_size_estimate() const { return _buf.size_bytes(); }
+
+uint32_t builder::restart_interval() const { return _restarts.capacity(); }
+
+void builder::add(internal::key key, iobuf&& value) {
+    if (_counter >= restart_interval()) {
```

Comment:
>REQUIRES: key must be lexicographically greater than the last key

Can we throw if this invariant is broken? Is the result of breaking the invariant some flavor of undefined behavior in a binary search down the line?